[5.7] Adding the ability to the container to resolve objects, through factory classes

ntavelis opened the pull request:
This PR adds the ability to the container to resolve objects, by using a factory class. The container will instantiate the factory class and resolve all its dependencies and then the container will call the method of the factory that was specified. The returned object from the `factoryClass@method` will get passed back.

Below follows an example that demonstrates the usage:

We register that when the class Car is required, the method `CarFactory@createRaceCar` will be called and the returned object from that method will be the object that will get passed back.

`app/Providers/AppServiceProvider.php`

```
/**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->bind(\App\Car::class, '\App\Factory\CarFactory@createRaceCar');
    }
```

The factory class will get instantiated from the container, and its dependencies will get resolved. We can use the factory to create  our object.

```
<?php

namespace App\Factory;

use App\Car;

class CarFactory
{
    public function createRaceCar(): Car
    {
        if(rand(0, 10) <= 5){
            return new Car('Tesla');
        }

        return new Car('Bentley');
    }
}
```

The class car:
```
<?php

namespace App;

class Car
{
    /**
     * @var string
     */
    private $model;

    /**
     * @param string $model
     */
    public function __construct(string $model)
    {

        $this->model = $model;
    }

    /**
     * @return string
     */
    public function getModel(): string
    {
        return 'I am a ' . $this->model;
    }
}
```
Also as mentioned above, the `Automatic Injection` is available for the factory's constructor and the method that is called.

```
<?php

namespace App\Factory;

use App\Car;
use App\User;
use Illuminate\Http\Request;

class CarFactory
{

    /**
     * @param Request $request
     */
    public function __construct(Request $request)
    {
        // the \Illuminate\Http\Request
        // gets injected into the factory's constructor
    }

    public function createRaceCar(User $user): Car
    {
        // The user class gets injected into the method
    }
}
```

Let me know what do you think. Thank you.

ntavelis committed:
Adding the ability to resolve objects from factories

ntavelis committed:
reverting the callClass static method to private, as it was

sisve commented:
This seems to lack any tests.

deleugpn commented:
What's the advantage of this over Closures?
How do you foresee the problem of binding a callable object as factory instead? When trying to bind an invokable class it will not work because the container will think we're binding a concrete instead of a factory. 

Maybe it would be better if this was a new public method instead. `$container->factory($class, callable $factory);`

+1 on sisve's comment.

ntavelis committed:
Adding tests and supporting additional syntax for factories binding

ntavelis committed:
Fixing the broken tests and the styling

ntavelis committed:
small style fix

ntavelis commented:
@deleugpn You have a fair point about the callables. The thing is that a closure is a callable also. That being said I changed the implementation to accept any valid callable(**but not closures**, we handle them as we used to) and valid sting that it is callable with the @ sign. 

So now the following syntax is supported:

```
// calling non static methods of a factory to resolve the concrete class, using the class@method syntax
$this->app->bind(\App\Car::class, '\App\Factory\CarFactory@createRaceCar');

//  calling non static methods of a factory to resolve the concrete class, using the array syntax
$factory = $this->app->make(\App\Factory\CarFactory::class);
$this->app->bind(\App\Car::class, [$factory, 'createRaceCar']);
```
```
// Using the array syntax to call a static method on a factory class to resolve the object
$this->app->bind(\App\Car::class, [$factory, 'staticallyCreateRaceCar']);
Using the :: syntax to call a static method on a factory class to resolve the object
$this->app->bind(\App\Car::class, [$factory, 'staticallyCreateRaceCar']);
```

Also added 4 tests to the container, for the above functionality.

deleugpn commented:
You're still missing my point. Imagine I want to do: `$app->bind(\My\Interface::class, \My\Concrete::class);`.
Now imagine that `Concrete` has a method `__invoke`. Your implementation will think that my `Concrete` is a factory, but in fact it isn't. It's just my concrete of my interface.

You will never know if a invokable is a concrete or a factory. This is my point about a separate public method.

ntavelis commented:
@deleugpn Hey man, sorry for not answering previously.

> You're still missing my point. Imagine I want to do: $app->bind(\My\Interface::class, \My\Concrete::class);.
Now imagine that Concrete has a method __invoke. Your implementation will think that my Concrete is a factory, but in fact it isn't. It's just my concrete of my interface.

> 
It is **NOT** true, this will work flawlessly, even if you have an __invoke() method, in your class.
`\My\Concrete::class` is a string, a string is never a callable, so this will work as always.

deleugpn commented:
I just checked https://3v4l.org/P9eEb and you are right 😲. I have been fooled by the smart Laravel Routing System for years https://github.com/laravel/framework/blob/5.7/src/Illuminate/Routing/RouteAction.php#L92

ntavelis commented:
Yeah, I don't know if a seperate method is needed. Anyway we will get more input on this and we will see!

ntavelis committed:
Adding a slash to the docblock

ntavelis commented on src/Illuminate/Container/Container.php:
```diff
@@ -257,6 +261,19 @@ protected function getClosure($abstract, $concrete)
+     * Get the Closure to be used when building a type from a factory.
+     * @param  string  $abstract
+     * @param  $concrete
+     * @return  Closure
```
added it

taylorotwell commented:
No plans to add this at the moment.
